### PR TITLE
ota-proxy enhancement 2

### PR DESCRIPTION
**Pull request by Bodong-Yang**
# Introduction
This PR aims to fix known and just-found bugs and improve the performance of ota-proxy.

# Bug fix
1. LRU cache implementation is not robust
2. Other bugs found during test...

# Changes
1. implement LRU cache only with db query.
2. implement simultaneously cache streaming.
3. optimize db performance by tuning db pragma settings.  
4. ota-proxy will only rotate cache if incoming entry is larger than 2KiB and smaller than 512MiB
5. other minor changes and modifications

# tickets
https://tier4.atlassian.net/browse/T4PB-17245
https://tier4.atlassian.net/browse/T4PB-16832

# TODO
- [x] local test
- [x] VM test(standalone)

### CODE REVIEW

**Comment by Bodong-Yang** [comment]
# Test result
with ~70,000 files, 1.9G in size
1. 1 client(1 thread), w/ ota-proxy: 180s~
2. 5 clients(1 thread each), w/ ota-proxy: 440s~, 130M~ mem usage

**Comment by Bodong-Yang** [comment]
Test needed to re-do as some changes are made.

**Comment by Bodong-Yang** [comment]
local test(multiple downloader) and VM test(standalone) passed.

**Comment by Bodong-Yang** [comment]
# Test report
### Test setup
1. 270265 files, 25G, local prepared image
### Test result
1. 5 downloader, w/ ota-proxy: 300M~, 1820s
2. 1 downloader, w/ ota-proxy: 60M~, 596s

File: ota_proxy/ota_cache.py [hunk 1]
```diff
@@ -563,13 +614,13 @@ def _background_check_free_space(self):
                     current_used_p >= cfg.DISK_USE_LIMIT_SOTF_P
```
Comment (obi-t4) [inline ota_proxy/ota_cache.py:614]:
```suggestion
                    current_used_p >= cfg.DISK_USE_LIMIT_SOFT_P
```

Comment (Bodong-Yang) [reply ota_proxy/ota_cache.py:614]: Oh typo!

File: ota_proxy/db.py [hunk 2]
```diff
@@ -105,84 +76,180 @@ def _connect_db(self, init: bool):
...
+def _proxy_cls_factory(cls, *, target):
+    for attr_n, attr in target.__dict__.items():
+        if not attr_n.startswith("_") and callable(attr):
+            # override method
+            setattr(cls, attr_n, _proxy_wrapper(attr_n))
+
+    return cls
+
+
+@partial(_proxy_cls_factory, target=OTACacheDB)
```
Comment (obi-t4) [inline ota_proxy/db.py:237]: @Bodong-Yang Don't you think this code is difficult? If there is no other implementation it is OK.

Comment (Bodong-Yang) [reply ota_proxy/db.py:237]: Emmm there are straight forward way to implement proxy class by coping-pasting every methods from the proxy target class, but that will make the code hard to be maintained(when original class is updated, the proxy class also needs to be updated).

So I use this generic way, which doesn't care about the details of the proxy target.

**Review by obi-t4 — APPROVED** [review]
Confirmed it works properly :+1:

**Pull request merged**